lightning/api_clnsocket.go

```diff
@@ -425,6 +429,259 @@ func (l *ClnSocketLightningAPI) GetNodeInfoFull(ctx context.Context, channels bo
+					FeeMsat:       ConvertAmount(one.FeeMsat),
+					IsSuccess:     success,
+					FailureString: one.FailReason,
+				})
+
+				if len(batch) >= int(batchSize) {
+					outChan <- batch
+					batch = make([]ForwardingEvent, 0, batchSize)
+				}
+			}
+
+			if len(batch) > 0 {
+				outChan <- batch
+				batch = make([]ForwardingEvent, 0, batchSize)
+			}
+
+			minTime = time.Time(reply.Entries[len(reply.Entries)-1].ReceivedTime).Add(-1 * time.Second)
+		}
+	}()
+
+	return outChan, errorChan
+}
+
+// GetInvoicesRaw - API call
+func (l *ClnSocketLightningAPI) GetInvoicesRaw(ctx context.Context, pendingOnly bool, pagination RawPagination) ([]RawMessage, *ResponseRawPagination, error) {
+	var (
+		reply   ClnRawInvoices
+		gettime ClnRawInvoiceTime
+	)
+
+	respPagination := &ResponseRawPagination{UseTimestamp: true}
+
+	err := l.CallWithTimeout(LISTINVOICES, []interface{}{}, &reply)
+	if err != nil {
+		return nil, respPagination, err
+	}
+
+	if err != nil {
+		return nil, respPagination, err
+	}
+
+	ret := make([]RawMessage, 0, len(reply.Entries))
+
+	minTime := time.Unix(1<<63-1, 0)
+	maxTime := time.Unix(0, 0)
+
+	for _, one := range reply.Entries {
+		err = json.Unmarshal(one, &gettime)
+		if err != nil {
+			return nil, respPagination, err
+		}
+
+		t := time.Unix(int64(gettime.Time), 0)
+
+		if t.Before(minTime) {
+			minTime = t
+		}
+		if t.After(maxTime) {
+			maxTime = t
+		}
+
+		m := RawMessage{
+			Implementation: l.Name,
+			Timestamp:      t,
+			Message:        one,
+		}
+
+		ret = append(ret, m)
+	}
+
+	respPagination.FirstTime = minTime
+	respPagination.LastTime = maxTime
+
+	return ret, respPagination, nil
+}
+
+// GetPaymentsRaw - API call
+func (l *ClnSocketLightningAPI) GetPaymentsRaw(ctx context.Context, includeIncomplete bool, pagination RawPagination) ([]RawMessage, *ResponseRawPagination, error) {
+	var (
+		reply   ClnRawPayments
+		gettime ClnRawPayTime
+	)
+
+	respPagination := &ResponseRawPagination{UseTimestamp: true}
+
+	err := l.CallWithTimeout(LISTPAYMENTS, []interface{}{}, &reply)
+	if err != nil {
+		return nil, respPagination, err
+	}
+
+	if err != nil {
+		return nil, respPagination, err
+	}
+
+	ret := make([]RawMessage, 0, len(reply.Entries))
+
+	minTime := time.Unix(1<<63-1, 0)
+	maxTime := time.Unix(0, 0)
+
+	for _, one := range reply.Entries {
+		err = json.Unmarshal(one, &gettime)
+		if err != nil {
+			return nil, respPagination, err
+		}
+
+		t := time.Unix(int64(gettime.Time), 0)
+
+		if t.Before(minTime) {
+			minTime = t
+		}
+		if t.After(maxTime) {
+			maxTime = t
+		}
+
+		m := RawMessage{
+			Implementation: l.Name,
+			Timestamp:      t,
+			Message:        one,
+		}
+
+		ret = append(ret, m)
+	}
+
+	respPagination.FirstTime = minTime
+	respPagination.LastTime = maxTime
+
+	return ret, respPagination, nil
+}
+
+// GetForwardsRaw - API call
+func (l *ClnSocketLightningAPI) GetForwardsRaw(ctx context.Context, pagination RawPagination) ([]RawMessage, *ResponseRawPagination, error) {
```

Comment:
most of the logic is duplicated for all types of payments. please extract common logic out and unit test it.